### DISCUSSION TITLE
NJU OS: Operating System Design and Implementation - CS自学指南

### DISCUSSION THREAD

**Discussion by giscus** [PKUFlyingPig/cs-self-learning#175] [Announcements]
# NJU OS: Operating System Design and Implementation - CS自学指南

CS自学指南

https://csdiy.wiki/%E6%93%8D%E4%BD%9C%E7%B3%BB%E7%BB%9F/NJUOS/

**Comment by 1475505** [comment]
前置课程是南大ICS的PA实验，jyy的B站空间有他2020年讲的习题课录像供参考。https://nju-projectn.github.io/ics-pa-gitbook/

**Reply by zixiangcode** [reply]
老哥，这是pa实验课，需不需要单独的再去看看“计算机系统基础”后再去看实验课啊？直接看实验课能听懂吗？操作系统这个课听了一节觉得好吃力www

**Reply by Tony-C11** [reply]
今年不用啦，jyy上课的时候说要把今年的课做成独立的一门

**Reply by kyangcmXF** [reply]
如果基础还不错的话 可以直接上手做PA PA的Lab0会介绍环境、生存指南。通过lab0可以初步检验是不是可以直接上手做PA 如果Lab0都看不懂 或者一知半解的话 那么应该补一下基础

**Reply by Cpt11** [reply]
推荐俩个便宜又好用的机场翻墙  平时如果只要google github 这些完全够用了
[一元机场](https://xn--4gq62f52gdss.com/#/register?code=APxjcVZh)   19元2年  100G/月

 [八方云](http://bafangyun.vip/#/register?code=0mobpNZJ)  12元/年  200G/月

**Reply by ghost** [reply]
有cs61c,csapp基础学2022年的是不是直接上手就可以了

**Reply by Zlatanwic** [reply]
2025jyy老师已经完全砍掉了oslab实验，只保留了minilab，有一些c语言基础和体系结构基础的应该都可以直接学习了

**Comment by heiyan-2020** [comment]
上了很多飞猪推荐的课，但没想到有一天自己在学校上过的课也能出现在这里。jyy确实是国内为数不多地让人眼前一亮的老师，他的课不光利用代码示例、实践等让学生对操作系统的理解不只是停留在概念上，更有很多方法论上的启发，绝对值得认真上完！
不过值得一提的是，课程的大实验（OSLab）才是真正的痛苦之源，难度可能很大哦（

**Reply by visonlee** [reply]
作为一个已经毕业多年的程序员, 也在蒋老师的课学习到许多

**Reply by CaoGongHui** [reply]
听视频还能听懂但是第一个实验就让实现一个tree命令,直接蒙了,毫无思路,要怎么学下去呢./(ㄒoㄒ)/~~

**Reply by yifeianyi** [reply]
回复楼上，如果觉得懵逼的话，可以先去做南大的PA，那是前置课程

**Reply by kyangcmXF** [reply]
看了操作系统的lab 才知道我们学校的lab属于玩具项目。虽然也是做操作系统子管理 只允许每个人独立完成 平均耗时一个月 不过真的差距蛮大的。
这里也不是贬低自己学校的lab 毕竟大学里有老师花时间设计lab 且能让学生花费一个月的小lab 也是花了心血的

**Reply by kyangcmXF** [reply]
如果觉得jyy设计的lab有难度的同学 可以看看实验材料 。如果把实验材料当作索引 按图索骥学习 （这实际上是科研自学的能力）也可以在一定程度上让自己飞速提升。然后再自己给自己设计lab

举个自己的例子：jyy在L5提到了自旋锁 我就用PA中提到的model checking（当然操作系统L4里也说到了）自己分别实现C && 内联汇编&& Java三个版本 然后分别用O2 O1 O0跑model checking 。一天时间就能入门:X86 lock section && memory order && memory coherence

**Reply by zhangshuchen206** [reply]
请问，南大的PA是纯编程吗？有没有需要测评的点？

**Reply by ZenusZhang** [reply]
请问lab0的框架代码是下载下来就可以编译的吗？
我编译选项ARCH=native可以正常运行，但是ARCH=x86-qemu就没有出现gui的框框。
我的环境是wsl2＋ubuntu。
我现在不太清楚这一个功能（在用qemu画黑白格）是需要自己实现的，还是我的环境有问题。
可以麻烦楼主回想一下么

**Comment by Zhang-Hanzhe** [comment]
JYYDS!

**Comment by Fontzs** [comment]
插眼

**Comment by lovelxc** [comment]
https://nju-projectn.github.io/ics-pa-gitbook/ics2021/index.html

**Comment by Delta-in-hub** [comment]
十分推荐,特别是实验,一定要做.

**Reply by jafox-12** [reply]
请问实验资源在哪找的呢👀

**Reply by kyangcmXF** [reply]
可以翻翻github 在jyy的gitbook里也谈到搜索的问题 建议先解决这个

**Reply by lsy1109** [reply]
jyy的gitbook指的是https://nju-projectn.github.io/ics-pa-gitbook/ics2021/index.html吗

**Reply by kyangcmXF** [reply]
实验资料看课程网站

**Comment by skyw4rd** [comment]
没有oj，怎么测试的呢。

**Reply by Infinite-Approximation** [reply]
老哥你现在知道如何进行完整的测试吗???

**Comment by eatvector** [comment]
有没有做了的相互测试

**Reply by jafox-12** [reply]
请问实验资源在哪找到的呢👀

**Comment by bartorat** [comment]
CSAPP的前五章可以替代这门课的前置课程吗？

**Reply by lovelxc** [reply]
可以

**Comment by eightyun** [comment]
为什么老师的个人课程网站进不去捏（哭）

**Reply by lovelxc** [reply]
> 为什么老师的个人课程网站进不去捏（哭）

http://jyywiki.cn/OS/2022/

**Comment by jafox-12** [comment]
请问实验资源在哪呢，没找到

**Comment by lsy1109** [comment]
请问njuos和mit的os 二选一的话 选谁比较好呢 他们内容上有什么区别吗

**Reply by lovelxc** [reply]
对新手来说，xv6是更好的选择，因为网上的博客写的很详细了。而nju的os虽好，但没有一定的工程能力是写不来的，不建议直接写。至于xv6的内容，可以在b站上找哈工深老师的简单入门介绍

**Reply by lsy1109** [reply]
我看介绍上说jyy老师是从很不同的角度讲os的，请问讲的内容和mit有什么差异的地方呢，独特在哪里呢？假设学完xv6还有学习njuos的必要吗？还是以后用到再学就好了阿？

**Reply by lovelxc** [reply]
jyy老师后面也会讲一下xv6，你可以只看课，然后试着写一写nju的lab，写不下去就只看课就好了，可以加q群1028681906看看大家怎么安排

**Reply by lyhellcat** [reply]
补一个HITsz的xv6入门介绍：https://www.bilibili.com/video/BV1Te4y1i77z/?spm_id_from=333.788.recommend_more_video.0

**Comment by KindofCrazy** [comment]
听说李樾和谭添老师的《软件分析》课程也很好（原谅我只是大一小白不知道具体讲的什么），大家有兴趣的可以在b站上看看

**Comment by kyangcmXF** [comment]
miniLab 和 osLab其实都可以在native运行 前提是修改makefile 将arch指向native 且视情况（其实就是make报错文本）来调整makefile和AM的源码 比方说禁止提交远端OJ 事实上校外同学没有token的话 也没必要提交。

但是提交给自己的私有仓库 做必要的源码管理是有意义的 ——jyy多次强调工程化管理 所以我们应该一开始学会用makefile

最后说一下没有OJ做miniLab && osLab还有意义吗？
有的 讲义中字里行间已经告诉我们怎么写单元测试（因为jyy将测试目的明确写在讲义上） 我们可以编写自己的测试用例 最起码 -m32 -m64 intptr_t 应该是知道的。
这样我们依然可以完成lab 且已经MS大多数同龄人了

-------------------分割线--------------------
感谢jyy && 博主 提供的材料 让我们飞速提升 夯实基础 :)

**Reply by LiguantingK** [reply]
您好，请问所有实验都可以本地做和自己做测试案例么？

**Reply by kyangcmXF** [reply]
可以的 至少我前四个lab都可以在本地native环境编译 编译需要修改makefile 配置必要的libxxx库 修改部分AM Klib代码（tip：测试用例记得回归这些修改过的.c .s文件 可能会炸）
因为是native环境 如果是从PA一路打怪过来 可以直接在NEMU里跑。当然直接在QEMU也是可以的

**Reply by LiguantingK** [reply]
十分感谢

**Comment by n-WN** [comment]
#### 真的是四颗星吗

**Comment by james-ma** [comment]
请问这个实验环境怎么搭建呢？ 看了课程的链接，上面有Abstractin Machine， 并没有说如何搭建实验环境

**Comment by james-ma** [comment]
看课程是用xv6做的core 代码， 但课程上没看到实验环境的搭建的指导， 是我漏掉了吗“？

**Comment by hdwang-sjtu** [comment]
有佬能简要说一下这个oslab和rcore的异同点嘛，比如实现的功能、难度啥的

**Comment by Dup4** [comment]
OS 2022 中的 M2:libco 在 OS 2023 中竟然移除了，是 蒋老师 觉得 libco 比较难吗，hh

**Reply by chenfs12138** [reply]
libco真的好，就是汇编那段黑魔法不好整，卡在这了

**Comment by Brightsx** [comment]
2023年的示例代码有一部分demo加载不出来了诶，有好心人有办法获得课堂演示的示例代码嘛？（第2、3、8、14、16、17、18课的加载不出来）

**Comment by ZenusZhang** [comment]
请问lab0的框架代码是下载下来就可以编译的吗？
我编译选项ARCH=native可以正常运行，但是ARCH=x86-qemu就没有出现gui的框框。
我的环境是wsl2＋ubuntu。
已经在这里卡一周了，没有进展，各位有什么解决方案吗？

**Comment by comi-zhang** [comment]
想问一下csapp可以算操作系统的一部分吗

**Comment by syccxdr** [comment]
最新的2024OS中jyy引用了csdiy.wiki中的前言！！

**Reply by Dup4** [reply]
好像没找到诶，能给指个路吗，是在 https://jyywiki.cn/OS/2024/ 里面吗

**Reply by wuuuiYan** [reply]
回复一楼：在第一节课“操作系统概述”的1.4小节如何学好操作系统的PPT里面

**Reply by PKUFlyingPig** [reply]
hhhh，开心，也算是jyy老师的老粉了

**Comment by Owenun** [comment]
找不到2023年代码了，有哪位大哥知道在哪有2023年代码吗

**Comment by phoulx** [comment]
TG讨论群(非官方): https://t.me/nju_os

**Comment by LRL52** [comment]
24 年 OS 课的课堂代码是没有公开吗？课上的代码有些展示不全，想看

**Reply by zixiangcode** [reply]
公开了，点击讲义页面的 vscode 的图标就能看见 wget 了

**Comment by ATreep** [comment]
已经学了 2/3 了，非常好的课程，不仅能学到操作系统的知识，重要的是还能了解到许多学习计算机科学的方法论（比如计算机中不存在魔法，一切都是有迹可循的；将程序看作是状态机，程序的执行看作是状态的转换等），个人非常建议每一个从事计算机专业的同学都看一下！
P.S. 课程的9个实验目前已经做了7个了，虽然对于本科生刚上手确实有难度，但是要相信自己能办到！每个实验完成了以后成就感都是无与伦比的～

**Reply by gulugulubing** [reply]
校外的是不是用不了OJ啊？影响大么

**Reply by ATreep** [reply]
> 校外的是不是用不了OJ啊？影响大么

是的用不了OJ，不过影响不是很大。
OJ也只是帮助检验实验代码正确性，其实每个实验的正确性规范在实验手册里都写得很清楚，自己自觉遵守就可以。对于pmm这类需要测试代码运行性能的实验，实验手册里测试代码也都给了出来，自己搭测试环境也是没问题的

**Reply by gulugulubing** [reply]
好的，谢谢。有机会再向你请教。

**Reply by gulugulubing** [reply]
那个M6的gpt作业你做了么，我运行 ./gpt 31373 612 结果和老师演示的一样，
但我运行 chat.py，输入 Ladies and 然后输出到NBA就停止了。而且我发现 31373 612并不是老师作业要求里说的Ladies and，而是hello there，是我装的tiktoken  有问题么？

**Reply by ATreep** [reply]
啊这个实验我还没做，抱歉

**Reply by gulugulubing** [reply]
请教一下，M2: 打印进程树，这个得到了一组[子进程,父进程]，然后要建立树，是不是得自己实现或者从外部导入一个hash表？

**Reply by z9nFoGJ** [reply]
你好 第一次尝试学这种网课 想问下课程链接附带的课程教材有必要搞一本实体书吗 我看是英文书有点害怕纯英文读不懂 又害怕没有教材学起来会不知所云Orz

**Comment by ClockworkBoat** [comment]
2023的课程代码是不维护了吧，am里面有点bug，可以把pa的am直接拉过来代替

**Comment by geng628** [comment]
推荐哪一年的版本呢？

**Comment by EnzoDing-rgb** [comment]
好东西

**Comment by WM20009** [comment]
现在蹭任何实验看起来都可以让ai在本地实现一份了😄甚至完全不需要官方代码框架
